Optimise long multiply + add/sub/neg on arm64.

c272 opened the pull request:
This PR optimises an extending multiply of two integers out to a long followed by an addition, subtraction or negation operation down to a single `[s/u]maddl`, `[s/u]subl` or `smnegl` instruction. It also adds a generic ternary operator which can be used in the future to support other three argument nodes.

The `spmidiff` results for this patch can be found at the below gist (performed on `win-arm64` since `linux-arm64` is currently broken as per #91257):
https://gist.github.com/c272/944721d2d083660e0ba7ec9fddb24de4

Minor regressions seen here are due to additional `mov` instructions changing the register allocation in cases where the add value is containable, preventing some previously performed `ldp` optimisations. Eliminating all cases of containable adds also removes many valid optimisations, so I have currently left it as is.

Contributes to https://github.com/dotnet/runtime/issues/68028

kunalspathak commented on src/coreclr/jit/codegenarm64.cpp:
```diff
@@ -2521,6 +2521,84 @@ void CodeGen::genCodeForMulHi(GenTreeOp* treeNode)
 }
 
+//------------------------------------------------------------------------
+// genCodeForMulAddSubLong: Generates code for int*int + long -> long.
```
also include an example for `SUB`.

kunalspathak reviewed (commented):
It seems adding `GenTreeTernOp` can be avoided. In the past, we directly recognized the pattern in codegen e.g. `madd`, `msub`. Did you try that approach and faced any problem?

a74nh commented:
> It seems adding `GenTreeTernOp` can be avoided. In the past, we directly recognized the pattern in codegen e.g. `madd`, `msub`. Did you try that approach and faced any problem?

I guess that would be possible, but it feels like it should be done during lowering (I have no overly strong opinion though).

My thought for `GenTreeTernOp` is that a follow on patch could move `SELECT`, `CMPXCHG` over to use it, which should simplify things a little.

tannergooding commented:
Rather than `GenTreeTernOp`, we already have `GenTreeHWIntrinsic` which supports multi-op and which is really oriented around these types of specialized instructions for both SIMD and Scalar code.

We can just introduce an `ArmBase_LongMultiplyAdd` or similar for internal use only and have lowering generate that node. The existing metadata table will make this very simple to support and likely require less overall code changes.

c272 committed:
Optimise long multiply + add/sub/neg on arm64.

This patch optimises an extending multiply of two integers
out to a long followed by an addition, subtraction or negation
operation down to a single [s/u]maddl, [s/u]subl or smnegl
instruction, represented as a GT_HWINTRINSIC.

c272 commented:
@tannergooding Thank you for the comment, I have reworked the patch to instead utilise `GT_HWINTRINSIC` instead of creating a new ternary node. The updated patch should also resolve the testing failures.

kunalspathak commented on src/coreclr/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -486,6 +486,14 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
             }
 
+            case NI_ArmBase_Arm64_MultiplyLongAdd:
+                ins = node->TypeIs(TYP_ULONG) ? INS_umaddl : INS_smaddl;
```
should this be `varTypeIsUnsigned(intrin.baseType)`?

kunalspathak commented on src/coreclr/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -1112,6 +1120,13 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+            case NI_ArmBase_Arm64_MultiplyLongAdd:
+            case NI_ArmBase_Arm64_MultiplyLongSub:
+                assert(opt == INS_OPTS_NONE);
+                GetEmitter()->emitIns_R_R_R_R(ins, EA_8BYTE, targetReg, op1Reg, op2Reg, op3Reg);
```
```suggestion
                GetEmitter()->emitIns_R_R_R_R(ins, emitSize, targetReg, op1Reg, op2Reg, op3Reg);
```

kunalspathak commented on src/coreclr/jit/lowerarmarch.cpp:
```diff
@@ -532,6 +532,16 @@ GenTree* Lowering::LowerBinaryArithmetic(GenTreeOp* binOp)
+
+        if (binOp->OperIs(GT_SUB))
+        {
+            // Attempt to optimise for smsubl.
```
```suggestion
            // Attempt to optimize for umsubl/smsubl.
```

kunalspathak commented on src/coreclr/jit/lowerarmarch.cpp:
```diff
@@ -2701,6 +2711,155 @@ void Lowering::TryLowerCselToCinc(GenTreeOp* select, GenTree* cond)
+
+//----------------------------------------------------------------------------------------------
+// TryLowerAddSubToCombinedMulOp: Attempt to convert ADD and SUB nodes to a combined multiply
+// and add/sub operation, extending. Conversion can only happen if the operands to the
```
```suggestion
// and add/sub operation. Conversion can only happen if the operands to the
```

kunalspathak commented on src/coreclr/jit/lowerarmarch.cpp:
```diff
@@ -2701,6 +2711,155 @@ void Lowering::TryLowerCselToCinc(GenTreeOp* select, GenTree* cond)
+    }
+
+    // Additional value must be of long size.
+    // The compiler will check the relevant signage for us, so we don't need to worry about that.
```
Not sure if I understood this comment. Can you please elaborate?

c272 replied:
Honestly, I'm also not sure what I was trying to get at here. It was probably to denote that we didn't need to check for `TYP_ULONG`, but I have removed the comment line.

kunalspathak reviewed (changes requested):
Need to resolve merge conflict and also fix the test failure:

```
Assert failure(PID 275 [0x00000113], Thread: 275 [0x0113]): Assertion failed 'compiler->compIsaSupportedDebugOnly(HWIntrinsicInfo::lookupIsa(intrin.id))' in 'System.Collections.Generic.Dictionary`2[System.__Canon,System.__Canon]:TryInsert(System.__Canon,System.__Canon,ubyte):ubyte:this' during 'Generate code' (IL size 668; hash 0x26ab038f; FullOpts)

    File: /__w/1/s/src/coreclr/jit/hwintrinsiccodegenarm64.cpp Line: 210
    Image: /root/helix/work/correlation/corerun
````

SingleAccretion commented on src/coreclr/jit/lowerarmarch.cpp:
```diff
@@ -2701,6 +2711,155 @@ void Lowering::TryLowerCselToCinc(GenTreeOp* select, GenTree* cond)
+    {
+        NamedIntrinsic intrinsicId =
+            op->OperIs(GT_ADD) ? NI_ArmBase_Arm64_MultiplyLongAdd : NI_ArmBase_Arm64_MultiplyLongSub;
+        GenTreeHWIntrinsic* outOp = comp->gtNewScalarHWIntrinsicNode(mul->IsUnsigned() ? TYP_ULONG : TYP_LONG,
```
> mul->IsUnsigned() ? TYP_ULONG : TYP_LONG

It is illegal to type nodes as `TYP_ULONG`.

SingleAccretion replied:
> add an assert to make it self-explanatory?

Definitely - there is a long-term desire to implement an "IR verifier" component (we already have a lot of the pieces, but they're a bit ad-hoc) that would check for things like that. It just hasn't been done yet.

c272 replied:
Thank you, wasn't aware of this. I've fixed it as `TYP_LONG` and used the intrinsic's base type to determine unsigned/signed.

c272 committed:
Formatting fixup.

Change-Id: Id0befd1a643620a1f4e0dcb0298f70cee296445d

c272 committed:
Disable generation on HW intrinsics disabled.

Change-Id: Iffcd54d6f2a3a25474e31f3a3cbf4072a8efcff8

tannergooding reviewed (approved):
This looks correct to me. It should get sign off from someone from @dotnet/jit-contrib before being merged

kunalspathak reviewed (approved):
LGTM
